### PR TITLE
 Export Framework as a stylized PDF from backend

### PR DESCRIPTION
**Pull request by anuradhas248**
It will Download Framework and it's all child items in stylized PDF format. This feature is implemented using PDF Office library. 
Story ID : KMS-3693.  It references git issues [#247](https://github.com/opensalt/opensalt/issues/247) and [#137](https://github.com/opensalt/opensalt/issues/137).

---
This change is [Reviewable](https://reviewable.io/reviews/opensalt/opensalt/447)

### CODE REVIEW

File: app/Resources/views/framework/doc_tree/export_pdf.html.twig [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<section>
+    <p style="line-height: 30px;"></p>
+    {% if pdfData.CFDocument.title is defined %}<p style="text-align:center"><h2>{{ pdfData.CFDocument.title }}</h2></p>{% endif %}
+{% if pdfData.CFDocument.publisher is defined %}<p style="text-align:center;"><h4>{{ pdfData.CFDocument.publisher }}</h4></p>{% endif %}
```
Comment (JosephLeon) [inline app/Resources/views/framework/doc_tree/export_pdf.html.twig]: Indent correctly.

File: tests/acceptance/features/framework/public_user/download_pdf.feature [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Feature: The framework can be downloaded in the form of pdf
+  In order to load a framework into another server
+  As an anonymous user
+  I need to download a pdf file of the framework
+
+  @incomplete @anonymous @pdf-file @7620-0007
+  Scenario: 0908-0000 An anonymous can see a framework
+    Given I am on a framework page
+    When I download the framework PDF file
+    Then I should see content in the PDF file
```
Comment (JosephLeon) [inline tests/acceptance/features/framework/public_user/download_pdf.feature]: Add newline at end of file.

File: tests/_support/Page/Framework.php [hunk 3]
```diff
@@ -1128,5 +1128,35 @@ public function iUploadTheItemTypeCASEFile()
...
+     */
+    public function iShouldSeeContentInThePDFFile(): Framework
+    {
+        $I = $this->I;
+
+        $PdfContent = $I->pdf2text($this->filename);
+        $I->assertNotEmpty($PdfContent, 'PDF file is empty');
+        $I->assertContains($I->getLastFrameworkTitle(), $PdfContent, 'Exported PDF');
+        return $this;
+    }
```
Comment (JosephLeon) [inline tests/_support/Page/Framework.php]: Remove extra empty lines.

File: composer.json [hunk 4]
```diff
@@ -91,7 +91,9 @@
 
         "matthimatiker/opcache-bundle": "^0.7",
         "qandidate/toggle-bundle": "^1.1.0",
-        "google/recaptcha": "~1.1"
+        "google/recaptcha": "~1.1",
+        "ggggino/wordbundle": "^1.0",
```
Comment (roverwolf) [inline composer.json]: I think this bundle is not really needed, it doesn't add much and makes things a bit more complicated.

File: composer.json [hunk 5]
```diff
@@ -91,7 +91,9 @@
...
+        "phpoffice/phpword": "v0.14.*"
```
Comment (roverwolf) [inline composer.json]: The "v" is not required and should be removed to be consistent with the rest of the file.

File: composer.json [hunk 6]
```diff
@@ -113,7 +115,8 @@
         "squizlabs/php_codesniffer": "^3.2",
         "escapestudios/symfony2-coding-standard": "^3.1",
         "friendsofphp/php-cs-fixer": "^2.10",
-        "ericmartel/codeception-email-mailtrap": "^1.0"
+        "ericmartel/codeception-email-mailtrap": "^1.0",
+        "tecnickcom/tcpdf": "6.*"
```
Comment (roverwolf) [inline composer.json]: It looks like this should be a regular requirement not a dev one.  Some deployments (like Heroku) will not install the dev requirements by default.

File: src/Controller/PdfExportController.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    use CommandDispatcherTrait;
+    /**
+     * @Route("/cfdoc/{id}/pdf", name="export_pdf_file")
+     * @Method("GET")
+     *
+     * @param int $id
+     */
+    public function exportPdfAction(int $id, Request $request)
+    {
+        $phpWordObject = $this->get('phpword')->createPHPWordObject();
```
Comment (roverwolf) [inline src/Controller/PdfExportController.php]: Should use dependency injection to the constructor or action over using `get` from the container.

But in this case, `new PhpWord()` is all that is really needed and is clearer.

File: src/Controller/PdfExportController.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            'framework/doc_tree/export_pdf.html.twig',
+            array(
+                    'pdfData' => json_decode($response->getContent(), true)
+                 )
+        );
+        Html::addHtml($section, htmlentities($html));
+        Settings::setPdfRendererName(Settings::PDF_RENDERER_TCPDF);
+        Settings::setPdfRendererPath('../vendor/tecnickcom/tcpdf');
+        $file = 'Framework.pdf';
+        $writer = $this->get('phpword')->createWriter($phpWordObject, 'PDF');
```
Comment (roverwolf) [inline src/Controller/PdfExportController.php]: Same here, just use the static function to create the writer.

File: src/Controller/PdfExportController.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        $response = new Response();
```
Comment (roverwolf) [inline src/Controller/PdfExportController.php]: Probably better to use a streamed response.  Check out the excel export action.

File: tests/_support/Helper/Guzzle.php [hunk 10]
```diff
@@ -147,4 +147,298 @@ public function _after(TestInterface $test)
             @unlink($file);
         }
     }
+
+    public function pdf2text($filename)
```
Comment (roverwolf) [inline tests/_support/Helper/Guzzle.php]: Probably better to move this (and the later methods) to a new helper for PDF parsing as it isn't related to guzzle.

File: src/Controller/PdfExportController.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public function exportPdfAction(int $id, Request $request)
+    {
+        $phpWordObject = $this->get('phpword')->createPHPWordObject();
+        // Create a new Page
+        $section = $phpWordObject->addSection();
+
+        $response = $this->forward('App\Controller\Framework\CfPackageController:exportAction', ['id' => $id, '_format' => 'json']);
+        $html = $this->renderView(
+            'framework/doc_tree/export_pdf.html.twig',
+            array(
```
Comment (roverwolf) [inline src/Controller/PdfExportController.php]: Prefer newer [] syntax over array().

File: tests/_support/Helper/Guzzle.php [hunk 12]
```diff
@@ -147,4 +147,298 @@ public function _after(TestInterface $test)
...
+    {
+    // Read the data from pdf file
+        $infile = @file_get_contents($filename, FILE_BINARY);
+        if (empty($infile))
```
Comment (roverwolf) [inline tests/_support/Helper/Guzzle.php]: Use {} around all if/else/for/etc. blocks.

File: app/Resources/views/framework/doc_tree/export_pdf.html.twig [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    <p style="line-height: 30px;"></p>
+    {% if pdfData.CFDocument.title is defined %}<p style="text-align:center"><h2>{{ pdfData.CFDocument.title }}</h2></p>{% endif %}
+    {% if pdfData.CFDocument.publisher is defined %}<p style="text-align:center;"><h4>{{ pdfData.CFDocument.publisher }}</h4></p>{% endif %}
+</section>
+
+<section>
+    <p style="line-height: 360px;"></p>
+    {% if pdfData.CFDocument.licenseURI.title is defined %}<p>Created by CfDoc Organization, under a {{ pdfData.CFDocument.licenseURI.title }} license</p>{% endif %}
+    {% if pdfData.CFDocument.statusStartDate is defined %}<p>Date : {{ pdfData.CFDocument.statusStartDate }}</p>{% endif %}
+    <p>Draft : Draft field is not present in case file right now</p>
```
Comment (roverwolf) [inline app/Resources/views/framework/doc_tree/export_pdf.html.twig]: adoptionStatus is the field to look at.  If empty then the default is draft.

**Comment by anuradhas248** [comment]
@roverwolf - I have updated code as per your suggestions. Please verify and confirm.

**Pull request merged**